getMeeting() in MeetingsSDKAdapter

arun3528 opened the pull request.

chrisadubois commented on src/MeetingsSDKAdapter.js:
```diff
@@ -0,0 +1,178 @@
+      id: meeting.Id,
+      title: meeting.meetingInfo.title,
+      startTime: '1:00', // check the values
+      endTime: 'not needed', // not sure why we need this
```
we do need this

chrisadubois commented on src/MeetingsSDKAdapter.js:
```diff
@@ -0,0 +1,178 @@
+   * @memberof MeetingsSDKAdapter
+   */
+  getMeeting(ID) {
+    const meeting = this.datasource.meetings.getMeetingByType('correlationId', ID);
```
lets change this from correlationId to id

chrisadubois commented on src/MeetingsSDKAdapter.js:
```diff
@@ -0,0 +1,178 @@
+    // TODO: what should the observable emit
+
+    meetingInstance.videoMuted = true;
+    this.getMeetingsObservables[meetingInstance.Id].next(meetingInstance);
```
don't we need to return something from the enum?

lalli-flores commented:
@arun3528, @chrisadubois are we gonna see an update on this?

chrisadubois commented:
hi @lalli-flores -- as mentioned we are in hackathon all this week and have not had time to make any updates. is the design for the super/base interface at a place for another review?

lalli-flores commented:
@chrisadubois, as we discussed before, we want smaller, separate PRs for the adapter, instead of one giant PR.
We've discussed `getMeeting` for the contents of the first PR. The interface for `getMeeting` has been defined and agreed on for weeks without a change. Not sure why the other parts of the interface would affect this PR.

adamweeks commented on src/MeetingsSDKAdapter.js:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    this.registerEvents();
+    this.syncMeetings();
```
Since this is just a one line function, we could just move the implementation into here.

adamweeks commented on src/MeetingsSDKAdapter.js:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  syncMeetings() {
+    this.datasource.meetings.syncMeetings();
```
We should probably await this since it returns a promise:
```suggestion
    await this.datasource.meetings.syncMeetings();
```

adamweeks commented on src/WebexSDKAdapter.js:
```diff
@@ -21,6 +23,7 @@ export default class WebexSDKAdapter {
   async connect() {
     await this.sdk.internal.device.register();
     await this.sdk.internal.mercury.connect();
+    this.meetingAdapter.connect();
```
```suggestion
    await this.meetingAdapter.connect();
```

adamweeks commented on src/WebexSDKAdapter.js:
```diff
@@ -29,5 +32,6 @@ export default class WebexSDKAdapter {
   async disconnect() {
     await this.sdk.internal.mercury.disconnect();
     await this.sdk.internal.device.unregister();
+    this.meetingAdapter.disconnect();
```
```suggestion
    await this.meetingAdapter.disconnect();
```

lalli-flores replied:
Is this defined anywhere? If you look at [CI test report](https://app.circleci.com/jobs/github/webex/sdk-component-adapter/360/parallel-runs/0/steps/0-104), integration tests are failing because `meeting.disconnect` is not defined

arun3528 replied:
added that

adamweeks commented on src/MeetingsSDKAdapter.js:
```diff
@@ -0,0 +1,102 @@
+   * @memberof MeetingsSDKAdapter
+   */
+  getMeeting(ID) {
+    const meeting = this.datasource.meetings.getMeetingByType('id', ID);
```
Since the meeting object returned from the plugin doesn't match the interface, you'll need to shape the object that gets returned a bit:

https://github.com/webex/component-adapter-interfaces/blob/master/src/MeetingsAdapter.js#L8

lalli-flores commented on src/MeetingsSDKAdapter.js:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  meetingRemoved(meetingInstance) {
+    this.meetingsObservables[meetingInstance.id].complet();
```
```suggestion
    this.meetingsObservables[meetingInstance.id].complete();
```

lalli-flores commented on src/MeetingsSDKAdapter.js:
```diff
@@ -0,0 +1,102 @@
+    // TODO: check if its correct
+    this.createObservable(meeting.id, meetingInstance);
+
+    this.registerMeetingsListener(meeting);
```
Is this defined somewhere?

arun3528 replied:
no removed it

lalli-flores commented on src/WebexSDKAdapter.js:
```diff
@@ -12,6 +13,7 @@ export default class WebexSDKAdapter {
   constructor(sdk) {
     this.peopleAdapter = new PeopleSDKAdapter(sdk);
     this.roomsAdapter = new RoomsSDKAdapter(sdk);
+    this.meetingAdapter = new MeetingSDKAdapter(sdk);
```
We try to keep properties alphabetical

lalli-flores commented on src/__mocks__/sdk.js:
```diff
@@ -14,6 +14,10 @@ export const mockSDKPerson = {
 };
 
+export const mockSDKMeeting = {
+  id: 'id',
```
```suggestion
  ID: 'id',
```
However, the shape of the meeting should match that of what is [defined in the interface](https://github.com/webex/component-adapter-interfaces/blob/master/src/MeetingsAdapter.js#L8). If the property has no value, then it should be set to `null`

lalli-flores commented on src/MeetingsSDKAdapter.js:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  registerEvents() {
+    this.datasource.meetings.on('meeting:added', this.meetingAdded);
```
Having events being fired outside the observable streams could cause unexpected race conditions. Events should be converted to observables with the [`fromEvent`](https://rxjs-dev.firebaseapp.com/api/index/function/fromEvent) operator. See how we handle events [`in other adapters`](https://github.com/webex/sdk-component-adapter/blob/master/src/RoomsSDKAdapter.js#L93)

lalli-flores commented on src/MeetignsSDKAdapter.test.js:
```diff
@@ -0,0 +1,42 @@
+    meetingSDKAdapter = new MeetingSDKAdapter(mockSDK);
+  });
+
+  describe('getMeeting() functionality', () => {
```
```suggestion
  describe('getMeeting()', () => {
```

lalli-flores commented on src/MeetignsSDKAdapter.test.js:
```diff
@@ -0,0 +1,42 @@
+      meetingSDKAdapter.getMeeting('id').subscribe((meeting) => {
+        expect(meeting).toEqual(
+          expect.objectContaining({
+            id: mockSDKMeeting.id,
```
```suggestion
            ID: mockSDKMeeting.id,
```

However, the shape of the meeting should match that of what is [defined in the interface](https://github.com/webex/component-adapter-interfaces/blob/master/src/MeetingsAdapter.js#L8). If the property has no value, then it should be set to `null`

lalli-flores commented on src/MeetingsSDKAdapter.js:
```diff
@@ -0,0 +1,102 @@
+
+  registerEvents() {
+    this.datasource.meetings.on('meeting:added', this.meetingAdded);
+    this.datasource.meetings.on('meeting:removed', this.meetingRemoved);
```
We should make event names constants, see [room adapter](https://github.com/webex/sdk-component-adapter/blob/master/src/RoomsSDKAdapter.js#L5)

arun3528 commented:
@lalli-flores  @adamweeks  i have fixed most of the review comments . I am still not clear on how to handle the create observable during an event and on getMeetings. Can you suggest some code there 

I am not sure how to convert

```
meeting.on('meeting:added',()=>{})
```

TO fromEvents

akoushke commented:
https://github.com/webex/sdk-component-adapter/pull/27 PR is a continuation  of this PR.
